domain/secretbackend/state/state.go

```diff
@@ -46,66 +47,253 @@ func NewState(factory coredatabase.TxnRunnerFactory, logger Logger) *State {
+			sb.TokenRotateInterval = database.NewNullDuration(*params.TokenRotateInterval)
+		}
+
+		err := tx.Query(ctx, upsertBackendStmt, sb).Run()
+		if database.IsErrConstraintUnique(err) {
+			return fmt.Errorf("%w: name %q", backenderrors.AlreadyExists, sb.Name)
+		}
+		if database.IsErrConstraintTrigger(err) {
+			return fmt.Errorf("%w: %q is immutable", backenderrors.Forbidden, sb.ID)
+		}
+		if err != nil {
+			return fmt.Errorf("cannot upsert secret backend %q: %w", params.Name, err)
+		}
+		if params.NextRotateTime != nil && !params.NextRotateTime.IsZero() {
+			err = tx.Query(ctx, upsertRotationStmt, SecretBackendRotation{
+				ID:               params.ID,
+				NextRotationTime: sql.NullTime{Time: *params.NextRotateTime, Valid: true},
+			}).Run()
+			if err != nil {
+				return fmt.Errorf("cannot upsert secret backend rotation time for %q: %w", params.Name, err)
+			}
+		}
+		if len(params.Config) == 0 {
+			return nil
+		}
+		namesToKeep := make(sqlair.S, 0, len(params.Config))
+		for k := range params.Config {
+			namesToKeep = append(namesToKeep, k)
+		}
+		err = tx.Query(ctx, clearConfigStmt, sqlair.M{"uuid": params.ID}, namesToKeep).Run()
+		if err != nil {
+			return fmt.Errorf("cannot clear secret backend config for %q: %w", params.Name, err)
+		}
+		for k, v := range params.Config {
+			// TODO: this needs to be fixed once the sqlair supports bulk insert.
+			err = tx.Query(ctx, upsertConfigStmt, SecretBackendConfig{
+				ID:      params.ID,
+				Name:    k,
+				Content: v.(string),
+			}).Run()
+			if err != nil {
+				return fmt.Errorf("cannot upsert secret backend config for %q: %w", params.Name, err)
+			}
+		}
+		return nil
 	})
 	return params.ID, domain.CoerceError(err)
 }
 
+func prepareDataForUpsertSecretBackend(
+	ctx context.Context, tx *sqlair.TX,
+	getStmt *sqlair.Statement,
+	params *secretbackend.UpsertSecretBackendParams,
+) error {
+	var existing SecretBackend
+	err := tx.Query(ctx, getStmt, sqlair.M{"uuid": params.ID}).Get(&existing)
+	if errors.Is(err, sqlair.ErrNoRows) {
+		// New insert.
+		if params.Name == "" {
+			return fmt.Errorf("%w: name is missing", backenderrors.NotValid)
+		}
+		if params.BackendType == "" {
+			return fmt.Errorf("%w: type is missing", backenderrors.NotValid)
+		}
+		return nil
+	} else if err != nil {
+		return errors.Trace(err)
+	}
+	// Update.
+	if existing.BackendType == "" {
+		return fmt.Errorf("backend type is empty for backend %q", params.ID)
+	}
+	if existing.Name == "" {
+		return fmt.Errorf("backend name is empty for backend %q", params.ID)
+	}
+	if params.BackendType != "" && params.BackendType != existing.BackendType {
+		// The secret backend type is immutable.
+		return fmt.Errorf(
+			"%w: cannot change backend type from %q to %q because backend type is immutable",
+			backenderrors.NotValid, existing.BackendType, params.BackendType,
+		)
+	}
+	// Fill in the existing backend type.
+	params.BackendType = existing.BackendType
+	if params.Name == "" {
```

Comment:
We've adopted the pattern that optional update params are pointers and set to nil if not used as part of the update.

Reply:
For update, sure it makes sense, but for insert, I kind of don't agree(because they are mandatory fields).
So it might be better to separate them.